File: week 4/points_and_segments/points_and_segments.py
# originally from http://www.cnblogs.com/dm1299/p/5571108.html

from itertools import chain

def fast_count_segments(starts, ends, points):
    cnt = [0] * len(points)
    a = zip(starts, [float('-inf')]*len(starts))
    b = zip(ends, [float('inf')]*len(ends))
    c = zip(points, range(len(points)))
    sortedlist = sorted(chain(a,b,c), key=lambda a : (a[0], a[1]))
    stack = []
    for i, j in sortedlist:
        if j == float('-inf'):
            stack.append(j)
        elif j == float('inf'):
            stack.pop()
        else: 
            cnt[j] = len(stack) 
    return cnt

def naive_count_segments(starts, ends, points):
    cnt = [0] * len(points)
    for i in range(len(points)):
        for j in range(len(starts)):
            if starts[j] <= points[i] <= ends[j]:
                cnt[i] += 1
    return cnt

if __name__ == '__main__':
    data = list(map(int, raw_input().split()))
    n = data[0]
    m = data[1]
    starts = [0] * n
    ends = [0] * n
    for i in range(n):
        starts[i], ends[i] = map(int, raw_input().split())
    points = list(map(int, raw_input().split()))
    #use fast_count_segments
    cnt = fast_count_segments(starts, ends, points)
    for x in cnt:
        print str(x)+ ' ',

File: week 4/points_and_segments/points_and_segments.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

vector<int> fast_count_segments(vector<pair<int,int> > &v,vector<int> points) {
  vector<int> cnt(points.size());
  //write your code here
  int n1=v.size(),n2=points.size(),i=0,j=0,k;
  for(i=0;i<n2;i++)
  {
    k=0;
    for(j=0;j<n1;j++)
    {
      if(points[i]>=v[j].first && points[i]<=v[j].second)
        k++;
      else if(points[i]<v[j].first)
        j=n1;
    }
    cnt[i] = k;
  }

  return cnt;
}
vector<int> naive_count_segments(vector<int> starts, vector<int> ends, vector<int> points) {
  vector<int> cnt(points.size());
  for (size_t i = 0; i < points.size(); i++) {
    for (size_t j = 0; j < starts.size(); j++) {  
      cnt[i] += starts[j] <= points[i] && points[i] <= ends[j];
    }
  }
  return cnt;
}
int main() {
  int n, m,a,b;
  std::cin >> n >> m;
  std::vector<pair<int,int> > v;
  for(int i =0;i<n;i++)
  {
    cin>>a>>b;
    v.push_back(make_pair(a,b));
  }
  vector<int> points(m);
  for (int i = 0; i < points.size(); i++) {
    std::cin >> points[i];
  }
  sort(v.begin(),v.end());
  vector<int> cnt = fast_count_segments(v, points);
  for (size_t i = 0; i < cnt.size(); i++) {
    std::cout << cnt[i] << ' ';
  }
}
